pkg/logs/tailer.go

```diff
@@ -0,0 +1,199 @@
+}
+
+// EventPayload is the mapping for fields in event payloads from request log tailing
+type EventPayload struct {
+	CreatedAt string `json:"created_at"`
+	Method    string `json:"method"`
+	RequestID string `json:"request_id"`
+	Status    int    `json:"status"`
+	URL       string `json:"url"`
+}
+
+// New creates a new Tailer
+func New(cfg *Config) *Tailer {
+	if cfg.Log == nil {
+		cfg.Log = &log.Logger{Out: ioutil.Discard}
+	}
+	return &Tailer{
+		cfg: cfg,
+		stripeAuthClient: stripeauth.NewClient(cfg.Key, &stripeauth.Config{
+			Log:        cfg.Log,
+			APIBaseURL: cfg.APIBaseURL,
+		}),
+		interruptCh: make(chan os.Signal, 1),
+	}
+}
+
+// Run sets the websocket connection
+func (tailer *Tailer) Run() error {
+	s := ansi.StartSpinner("Getting ready...", tailer.cfg.Log.Out)
+
+	// Intercept Ctrl+c so we can do some clean up
+	signal.Notify(tailer.interruptCh, os.Interrupt, syscall.SIGTERM)
+
+	filters, err := jsonifyFilters(tailer.cfg.Filters)
+	if err != nil {
+		tailer.cfg.Log.Fatalf("Error while converting log filters to JSON encoding: %v", err)
+	}
+
+	session, err := tailer.stripeAuthClient.Authorize(tailer.cfg.DeviceName, tailer.cfg.WebSocketFeature, &filters)
+	if err != nil {
+		// TODO: better error handling / retries
+		tailer.cfg.Log.Fatalf("Error while authenticating with Stripe: %v", err)
+	}
+
+	tailer.webSocketClient = websocket.NewClient(
+		session.WebSocketURL,
+		session.WebSocketID,
+		session.WebSocketAuthorizedFeature,
+		&websocket.Config{
+			EventHandler:      websocket.EventHandlerFunc(tailer.processRequestLogEvent),
+			Log:               tailer.cfg.Log,
+			NoWSS:             tailer.cfg.NoWSS,
+			ReconnectInterval: time.Duration(session.ReconnectDelay) * time.Second,
+		},
+	)
+	go tailer.webSocketClient.Run()
+
+	ansi.StopSpinner(s, "Ready! You're now waiting to receive API request logs (^C to quit)", tailer.cfg.Log.Out)
+
+	// Block until Ctrl+C is received
+	<-tailer.interruptCh
+
+	log.WithFields(log.Fields{
+		"prefix": "logs.Tailer.Run",
+	}).Debug("Ctrl+C received, cleaning up...")
+
+	if tailer.webSocketClient != nil {
+		tailer.webSocketClient.Stop()
+	}
+
+	log.WithFields(log.Fields{
+		"prefix": "logs.Tailer.Run",
+	}).Debug("Bye!")
+
+	return nil
+}
+
+func (tailer *Tailer) processRequestLogEvent(msg websocket.IncomingMessage) {
+	if msg.RequestLogEvent == nil {
+		tailer.cfg.Log.Warn("WebSocket specified for request logs received non-request-logs event")
+		return
+	}
+
+	requestLogEvent := msg.RequestLogEvent
+
+	tailer.cfg.Log.WithFields(log.Fields{
+		"prefix":     "logs.Tailer.processRequestLogEvent",
+		"webhook_id": requestLogEvent.RequestLogID,
+	}).Debugf("Processing request log event")
+
+	var payload EventPayload
+	if err := json.Unmarshal([]byte(requestLogEvent.EventPayload), &payload); err != nil {
+		tailer.cfg.Log.Warn("Received malformed payload: ", err)
+	}
+
+	// Don't show stripecli/sessions logs since they're generated by the CLI
+	if payload.URL == "/v1/stripecli/sessions" {
```

Comment:
What would you think if we added a super verbose option that prints these out too? Doesn't need to happen now though.

Reply:
We could...but I don't know if ever printing `/v1/stripecli/sessions` logs is useful for the user, since they have no way of triggering it themselves and have no idea it even exists -- it doesn't show up in their dashboard logs I believe.